//
// DO NOT EDIT
//
// Generated by swift-atproto
//

import SwiftAtproto
import Foundation

extension comatprototypes {
    public enum SyncGetRecord_Error: XRPCError {
        case recordnotfound(String?)
        case repodeactivated(String?)
        case reponotfound(String?)
        case reposuspended(String?)
        case repotakendown(String?)
        case unexpected(error: String?, message: String?)

        public init(error: UnExpectedError) {
            switch error.error {
            case "RecordNotFound":
                self = .recordnotfound(error.message)
            case "RepoNotFound":
                self = .reponotfound(error.message)
            case "RepoTakendown":
                self = .repotakendown(error.message)
            case "RepoSuspended":
                self = .reposuspended(error.message)
            case "RepoDeactivated":
                self = .repodeactivated(error.message)
            default:
                self = .unexpected(error: error.error, message: error.message)
            }
        }

        public var error: String? {
            switch self {
            case .recordnotfound:
                return "RecordNotFound"
            case .reponotfound:
                return "RepoNotFound"
            case .repotakendown:
                return "RepoTakendown"
            case .reposuspended:
                return "RepoSuspended"
            case .repodeactivated:
                return "RepoDeactivated"
            case let .unexpected(error, _):
                return error
            }
        }

        public var message: String? {
            switch self {
            case let .recordnotfound(message):
                return message
            case let .reponotfound(message):
                return message
            case let .repotakendown(message):
                return message
            case let .reposuspended(message):
                return message
            case let .repodeactivated(message):
                return message
            case let .unexpected(_, message):
                return message
            }
        }
    }

    public static func SyncGetRecord(client: any XRPCClientProtocol, collection: String, commit: String?, did: String, rkey: String) async throws -> Data {
        let params: Parameters = ["collection": .string(collection), "commit": .string(commit), "did": .string(did), "rkey": .string(rkey)]
        do {
            return try await client.fetch(endpoint: "com.atproto.sync.getRecord", contentType: "*/*", httpMethod: .get, params: params, input: Bool?.none, retry: true)
        } catch let error as UnExpectedError {
            throw SyncGetRecord_Error(error: error)
        }
    }
}
